org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageServerWrapper.java

```diff
@@ -553,46 +575,39 @@ private void shutdown() {
 		this.serverCapabilities = null;
 		this.dynamicRegistrations.clear();
 
-		final Future<?> serverFuture = this.launcherFuture;
-		final StreamConnectionProvider provider = this.lspStreamProvider;
-		final LanguageServer languageServerInstance = this.languageServer;
 		ResourcesPlugin.getWorkspace().removeResourceChangeListener(workspaceFolderUpdater);
 
-		Runnable shutdownKillAndStopFutureAndProvider = () -> {
-			if (languageServerInstance != null) {
-				CompletableFuture<Object> shutdown = languageServerInstance.shutdown();
-				try {
-					shutdown.get(5, TimeUnit.SECONDS);
-				} catch (InterruptedException ex) {
-					Thread.currentThread().interrupt();
-				} catch (Exception ex) {
-					LanguageServerPlugin.logError(ex.getClass().getSimpleName() + " occurred during shutdown of " + languageServerInstance, ex); //$NON-NLS-1$
+		CompletableFuture.runAsync(() -> {
+			synchronized(workingContext) {
+				if (workingContext.languageServer != null) {
+					CompletableFuture<Object> shutdown = workingContext.languageServer.shutdown();
+					try {
+						shutdown.get(5, TimeUnit.SECONDS);
+					} catch (InterruptedException ex) {
+						Thread.currentThread().interrupt();
+					} catch (Exception ex) {
+						LanguageServerPlugin.logError(ex.getClass().getSimpleName() + " occurred during shutdown of " //$NON-NLS-1$
+								+ workingContext.languageServer, ex);
+					}
 				}
-			}
 
-			if (serverFuture != null) {
-				serverFuture.cancel(true);
-			}
+				if (workingContext.launcherFuture != null) {
```

Comment:
Now that we have an object that represents a working context, could we maybe refactor this bit by adding a method `close` (or similar name) and move the `cancel` of the future, the `exit` on the server and the `stop` of the streamProvider into it?